File: core-clients/python/ramen_ai/governed.py
"""Isolated JSON and SSE transport for governed generation."""

from __future__ import annotations

import json
from collections.abc import Iterator, Mapping, Sequence
from typing import Any, Literal, cast

import httpx

from .governed_errors import (
    GovernanceDeniedException,
    GovernedGenerationException,
)
from .governed_types import (
    GovernedAccounting,
    GovernedAttemptMetadata,
    GovernedBlockedData,
    GovernedCompleteData,
    GovernedCompleteEnvelope,
    GovernedCompleteEvent,
    GovernedEvaluationSummary,
    GovernedGenerationOptions,
    GovernedHeartbeatData,
    GovernedHeartbeatEvent,
    GovernedProviderName,
    GovernedStatusData,
    GovernedStatusEvent,
    GovernedStatusStage,
    GovernedStreamEvent,
    GovernedTokenUsage,
)

_GOVERNED_PATH = "/api/v1/generate/governed"
_STATUS_STAGES = {"accepted", "generating", "evaluating", "regenerating"}


def generate_governed(
    client: httpx.Client,
    prompt: str,
    *,
    policy_ids: Sequence[str] | None = None,
    bundle_ids: Sequence[str] | None = None,
    max_retries: Literal[0, 1] = 1,
    generation: GovernedGenerationOptions | None = None,
    provider_key: str | None = None,
    provider_name: GovernedProviderName | None = None,
) -> GovernedCompleteData:
    """Execute a governed generation request and return released content."""

    body = _build_body(prompt, policy_ids, bundle_ids, max_retries, generation)
    headers = _provider_headers(provider_key, provider_name)
    try:
        response = client.post(
            _GOVERNED_PATH,
            json=body,
            headers=headers or None,
        )
    except httpx.TimeoutException as error:
        raise GovernedGenerationException(
            None, "TRANSPORT_TIMEOUT", "Governed generation timed out"
        ) from error
    except httpx.HTTPError as error:
        raise GovernedGenerationException(
            None, "TRANSPORT_ERROR", "Governed generation request failed"
        ) from error

    payload = _decode_response_json(response)
    if not response.is_success or payload.get("success") is not True:
        _raise_api_error(response.status_code, payload)
    return _parse_complete_data(payload.get("data"), response.status_code)


def generate_governed_stream(
    client: httpx.Client,
    prompt: str,
    *,
    policy_ids: Sequence[str] | None = None,
    bundle_ids: Sequence[str] | None = None,
    max_retries: Literal[0, 1] = 1,
    generation: GovernedGenerationOptions | None = None,
    provider_key: str | None = None,
    provider_name: GovernedProviderName | None = None,
) -> Iterator[GovernedStreamEvent]:
    """Yield governed status events and the successful terminal event."""

    body = _build_body(prompt, policy_ids, bundle_ids, max_retries, generation)
    headers = {
        "Accept": "text/event-stream",
        **_provider_headers(provider_key, provider_name),
    }

    try:
        with client.stream(
            "POST",
            _GOVERNED_PATH,
            json=body,
            headers=headers,
        ) as response:
            if not response.is_success:
                response.read()
                payload = _decode_response_json(response)
                _raise_api_error(response.status_code, payload)

            content_type = response.headers.get("Content-Type", "").lower()
            if not content_type.startswith("text/event-stream"):
                raise GovernedGenerationException(
                    response.status_code,
                    "STREAM_PROTOCOL_ERROR",
                    "Governed generation did not return an SSE stream",
                )

            event_name: str | None = None
            data_lines: list[str] = []
            terminal_seen = False

            for line in response.iter_lines():
                if line == "":
                    parsed = _dispatch_sse_event(
                        event_name, data_lines, response.status_code
                    )
                    event_name = None
                    data_lines = []
                    if parsed is None:
                        continue
                    yield parsed
                    if parsed.event == "complete":
                        terminal_seen = True
                        return
                    continue

                if line.startswith(":"):
                    continue
                field, separator, value = line.partition(":")
                if not separator:
                    value = ""
                elif value.startswith(" "):
                    value = value[1:]
                if field == "event":
                    event_name = value
                elif field == "data":
                    data_lines.append(value)

            if event_name is not None or data_lines:
                parsed = _dispatch_sse_event(
                    event_name, data_lines, response.status_code
                )
                if parsed is not None:
                    yield parsed
                    if parsed.event == "complete":
                        terminal_seen = True
                        return

            if not terminal_seen:
                raise GovernedGenerationException(
                    response.status_code,
                    "STREAM_TERMINATED",
                    "Governed generation stream ended without a terminal event",
                )
    except GovernedGenerationException:
        raise
    except httpx.TimeoutException as error:
        raise GovernedGenerationException(
            None, "TRANSPORT_TIMEOUT", "Governed generation stream timed out"
        ) from error
    except httpx.HTTPError as error:
        raise GovernedGenerationException(
            None, "TRANSPORT_ERROR", "Governed generation stream failed"
        ) from error


def _build_body(
    prompt: str,
    policy_ids: Sequence[str] | None,
    bundle_ids: Sequence[str] | None,
    max_retries: Literal[0, 1],
    generation: GovernedGenerationOptions | None,
) -> dict[str, Any]:
    if not isinstance(prompt, str) or not prompt.strip() or len(prompt) > 10_000:
        raise ValueError("prompt must be a non-blank string of at most 10,000 characters")
    if not policy_ids and not bundle_ids:
        raise ValueError("Provide at least one of 'bundle_ids' or 'policy_ids'.")
    if max_retries not in (0, 1):
        raise ValueError("max_retries must be 0 or 1")

    body: dict[str, Any] = {"prompt": prompt, "max_retries": max_retries}
    if policy_ids:
        body["policy_ids"] = list(policy_ids)
    if bundle_ids:
        body["bundle_ids"] = list(bundle_ids)
    if generation is not None:
        generation_body: dict[str, int | float] = {}
        if generation.temperature is not None:
            if not 0 <= generation.temperature <= 2:
                raise ValueError("generation.temperature must be between 0 and 2")
            generation_body["temperature"] = generation.temperature
        if generation.max_tokens is not None:
            if not 1 <= generation.max_tokens <= 4096:
                raise ValueError("generation.max_tokens must be between 1 and 4096")
            generation_body["max_tokens"] = generation.max_tokens
        body["generation"] = generation_body
    return body


def _provider_headers(
    provider_key: str | None,
    provider_name: GovernedProviderName | None,
) -> dict[str, str]:
    if not provider_key:
        return {}
    headers = {"X-Provider-Key": provider_key}
    if provider_name:
        headers["X-Provider"] = provider_name
    return headers


def _decode_response_json(response: httpx.Response) -> Mapping[str, Any]:
    try:
        payload = response.json()
    except (json.JSONDecodeError, UnicodeDecodeError, ValueError) as error:
        raise GovernedGenerationException(
            response.status_code,
            "PARSE_ERROR",
            "Failed to parse governed generation response",
        ) from error
    if not isinstance(payload, Mapping):
        raise GovernedGenerationException(
            response.status_code,
            "PARSE_ERROR",
            "Governed generation response must be a JSON object",
        )
    return cast(Mapping[str, Any], payload)


def _raise_api_error(status: int, payload: Mapping[str, Any]) -> None:
    error_value = payload.get("error")
    error = error_value if isinstance(error_value, Mapping) else {}
    code = error.get("code") if isinstance(error.get("code"), str) else "UNKNOWN_ERROR"
    message = (
        error.get("message")
        if isinstance(error.get("message"), str)
        else f"Governed generation failed with HTTP {status}"
    )
    details = error.get("details", payload.get("data"))
    if status == 422 and code == "GOVERNED_OUTPUT_BLOCKED":
        data = _parse_blocked_data(payload.get("data"), status)
        raise GovernanceDeniedException(message, data)
    raise GovernedGenerationException(status, code, message, details)


def _dispatch_sse_event(
    event_name: str | None,
    data_lines: list[str],
    status: int,
) -> GovernedStreamEvent | None:
    if event_name not in {"status", "heartbeat", "complete", "blocked", "error"}:
        return None
    if not data_lines:
        raise GovernedGenerationException(
            status, "STREAM_PARSE_ERROR", f"SSE event {event_name} has no data"
        )
    try:
        payload = json.loads("\n".join(data_lines))
    except json.JSONDecodeError as error:
        raise GovernedGenerationException(
            status,
            "STREAM_PARSE_ERROR",
            f"Failed to parse governed generation SSE event: {event_name}",
        ) from error
    if not isinstance(payload, Mapping):
        raise GovernedGenerationException(
            status,
            "STREAM_PARSE_ERROR",
            f"Governed generation SSE event {event_name} must contain an object",
        )

    try:
        if event_name == "status":
            stage_value = _required(payload, "stage", str)
            if stage_value not in _STATUS_STAGES:
                raise ValueError("invalid status stage")
            violations_value = payload.get("violations")
            violations = (
                _expect_int(violations_value, "violations")
                if violations_value is not None
                else None
            )
            return GovernedStatusEvent(
                event="status",
                data=GovernedStatusData(
                    stage=cast(GovernedStatusStage, stage_value),
                    attempt=_required_int(payload, "attempt"),
                    violations=violations,
                ),
            )
        if event_name == "heartbeat":
            return GovernedHeartbeatEvent(
                event="heartbeat",
                data=GovernedHeartbeatData(
                    timestamp=_required(payload, "timestamp", str)
                ),
            )
        if event_name == "complete":
            if payload.get("success") is not True:
                raise ValueError("invalid complete terminal event")
            return GovernedCompleteEvent(
                event="complete",
                data=GovernedCompleteEnvelope(
                    success=True,
                    data=_parse_complete_data(payload.get("data"), status),
                ),
            )

        error_value = payload.get("error")
        error = _expect_mapping(error_value, "error")
        code = _required(error, "code", str)
        message = _required(error, "message", str)
        logical_status = _required_int(error, "http_status")
        if event_name == "blocked":
            if code != "GOVERNED_OUTPUT_BLOCKED" or logical_status != 422:
                raise ValueError("invalid blocked terminal event")
            raise GovernanceDeniedException(
                message, _parse_blocked_data(payload.get("data"), logical_status)
            )
        raise GovernedGenerationException(
            logical_status, code, message, payload.get("data")
        )
    except GovernedGenerationException:
        raise
    except (KeyError, TypeError, ValueError) as error:
        raise GovernedGenerationException(
            status,
            "STREAM_PARSE_ERROR",
            f"Invalid governed generation SSE event: {event_name}",
        ) from error


def _parse_complete_data(value: Any, status: int) -> GovernedCompleteData:
    try:
        data = _expect_mapping(value, "data")
        usage_value = data.get("usage")
        return GovernedCompleteData(
            content=_required(data, "content", str),
            provider=_required(data, "provider", str),
            model=_required(data, "model", str),
            attempts=_required_int(data, "attempts"),
            attempt_metadata=_parse_attempt_metadata(data.get("attempt_metadata")),
            evaluation=_parse_evaluation(data.get("evaluation")),
            accounting=_parse_accounting(data.get("accounting")),
            total_duration_ms=_required_int(data, "total_duration_ms"),
            usage=_parse_usage(usage_value) if usage_value is not None else None,
        )
    except GovernedGenerationException:
        raise
    except (KeyError, TypeError, ValueError) as error:
        raise GovernedGenerationException(
            status, "PARSE_ERROR", "Invalid governed generation completion payload"
        ) from error


def _parse_blocked_data(value: Any, status: int) -> GovernedBlockedData:
    try:
        data = _expect_mapping(value, "data")
        return GovernedBlockedData(
            attempts=_required_int(data, "attempts"),
            attempt_metadata=_parse_attempt_metadata(data.get("attempt_metadata")),
            evaluation=_parse_evaluation(data.get("evaluation")),
            accounting=_parse_accounting(data.get("accounting")),
            total_duration_ms=_required_int(data, "total_duration_ms"),
        )
    except GovernedGenerationException:
        raise
    except (KeyError, TypeError, ValueError) as error:
        raise GovernedGenerationException(
            status, "PARSE_ERROR", "Invalid governed generation blocked payload"
        ) from error


def _parse_usage(value: Any) -> GovernedTokenUsage:
    data = _expect_mapping(value, "usage")
    return GovernedTokenUsage(
        prompt_tokens=_required_int(data, "prompt_tokens"),
        completion_tokens=_required_int(data, "completion_tokens"),
        total_tokens=_required_int(data, "total_tokens"),
    )


def _parse_attempt_metadata(value: Any) -> tuple[GovernedAttemptMetadata, ...]:
    items = _expect_list(value, "attempt_metadata")
    parsed: list[GovernedAttemptMetadata] = []
    for item in items:
        data = _expect_mapping(item, "attempt_metadata item")
        usage_value = data.get("usage")
        parsed.append(
            GovernedAttemptMetadata(
                attempt=_required_int(data, "attempt"),
                provider=_required(data, "provider", str),
                model=_required(data, "model", str),
                generation_duration_ms=_required_int(
                    data, "generation_duration_ms"
                ),
                evaluation_duration_ms=_required_int(
                    data, "evaluation_duration_ms"
                ),
                policies_evaluated=_required_int(data, "policies_evaluated"),
                allowed=_required_bool(data, "allowed"),
                usage=_parse_usage(usage_value) if usage_value is not None else None,
            )
        )
    return tuple(parsed)


def _parse_accounting(value: Any) -> GovernedAccounting:
    data = _expect_mapping(value, "accounting")
    return GovernedAccounting(
        generation_attempts=_required_int(data, "generation_attempts"),
        evaluation_batches=_required_int(data, "evaluation_batches"),
        policy_evaluations=_required_int(data, "policy_evaluations"),
    )


def _parse_evaluation(value: Any) -> GovernedEvaluationSummary:
    data = _expect_mapping(value, "evaluation")
    receipt_id = data.get("receipt_id")
    if receipt_id is not None and not isinstance(receipt_id, str):
        raise TypeError("evaluation.receipt_id must be a string")
    return GovernedEvaluationSummary(
        allowed=_required_bool(data, "allowed"),
        policy_ids=tuple(_expect_string_list(data.get("policy_ids"), "policy_ids")),
        policies_evaluated=_required_int(data, "policies_evaluated"),
        policies_passed=_required_int(data, "policies_passed"),
        policies_failed=_required_int(data, "policies_failed"),
        policies_errored=_required_int(data, "policies_errored"),
        violation_count=_required_int(data, "violation_count"),
        statutory_anchors=tuple(
            _expect_string_list(data.get("statutory_anchors"), "statutory_anchors")
        ),
        receipt_id=receipt_id,
    )


def _expect_mapping(value: Any, name: str) -> Mapping[str, Any]:
    if not isinstance(value, Mapping):
        raise TypeError(f"{name} must be an object")
    return cast(Mapping[str, Any], value)


def _expect_list(value: Any, name: str) -> list[Any]:
    if not isinstance(value, list):
        raise TypeError(f"{name} must be an array")
    return value


def _expect_string_list(value: Any, name: str) -> list[str]:
    items = _expect_list(value, name)
    if any(not isinstance(item, str) for item in items):
        raise TypeError(f"{name} must contain strings")
    return cast(list[str], items)


def _required(mapping: Mapping[str, Any], key: str, expected: type[Any]) -> Any:
    value = mapping[key]
    if not isinstance(value, expected):
        raise TypeError(f"{key} has an invalid type")
    return value


def _expect_int(value: Any, name: str) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise TypeError(f"{name} must be an integer")
    return value


def _required_int(mapping: Mapping[str, Any], key: str) -> int:
    return _expect_int(mapping[key], key)


def _required_bool(mapping: Mapping[str, Any], key: str) -> bool:
    value = mapping[key]
    if not isinstance(value, bool):
        raise TypeError(f"{key} must be a boolean")
    return value

File: core-clients/node/src/governed.ts
import {
  GovernanceDeniedException,
  GovernedGenerationException,
} from "./governed-errors.js";
import type {
  GenerateGovernedOptions,
  GovernedAccounting,
  GovernedAttemptMetadata,
  GovernedBlockedData,
  GovernedCompleteData,
  GovernedCompleteEvent,
  GovernedEvaluationSummary,
  GovernedGenerateWireRequest,
  GovernedHeartbeatEvent,
  GovernedStatusEvent,
  GovernedStatusStage,
  GovernedStreamEvent,
  GovernedTokenUsage,
} from "./governed-types.js";

const GOVERNED_PATH = "/api/v1/generate/governed";
const GOVERNED_TIMEOUT_MS = 60_000;
const STATUS_STAGES = new Set<GovernedStatusStage>([
  "accepted",
  "generating",
  "evaluating",
  "regenerating",
]);

type UnknownRecord = Record<string, unknown>;

export interface GovernedTransportConfig {
  apiKey: string;
  baseUrl: string;
  providerKey?: string;
  providerName?: string;
  fetchImpl: typeof fetch;
}

export async function generateGoverned(
  config: GovernedTransportConfig,
  prompt: string,
  options: GenerateGovernedOptions,
): Promise<GovernedCompleteData> {
  const body = buildBody(prompt, options);
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), GOVERNED_TIMEOUT_MS);

  try {
    const response = await config.fetchImpl(`${config.baseUrl}${GOVERNED_PATH}`, {
      method: "POST",
      headers: buildHeaders(config, "application/json"),
      body: JSON.stringify(body),
      signal: controller.signal,
    });
    const payload = await decodeJson(response);
    if (!response.ok || payload.success !== true) {
      throwApiError(response.status, payload);
    }
    return parseCompleteData(payload.data, response.status);
  } catch (error) {
    if (error instanceof GovernedGenerationException) throw error;
    if (controller.signal.aborted) {
      throw new GovernedGenerationException(
        null,
        "TRANSPORT_TIMEOUT",
        "Governed generation timed out",
      );
    }
    throw new GovernedGenerationException(
      null,
      "TRANSPORT_ERROR",
      "Governed generation request failed",
      error,
    );
  } finally {
    clearTimeout(timer);
  }
}

export async function* generateGovernedStream(
  config: GovernedTransportConfig,
  prompt: string,
  options: GenerateGovernedOptions,
): AsyncGenerator<GovernedStreamEvent> {
  const body = buildBody(prompt, options);
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), GOVERNED_TIMEOUT_MS);
  let reader: ReadableStreamDefaultReader<Uint8Array> | undefined;
  let terminalSeen = false;

  try {
    const response = await config.fetchImpl(`${config.baseUrl}${GOVERNED_PATH}`, {
      method: "POST",
      headers: buildHeaders(config, "text/event-stream"),
      body: JSON.stringify(body),
      signal: controller.signal,
    });

    if (!response.ok) {
      const payload = await decodeJson(response);
      throwApiError(response.status, payload);
    }

    const contentType = response.headers.get("Content-Type")?.toLowerCase() ?? "";
    if (!contentType.startsWith("text/event-stream") || !response.body) {
      throw new GovernedGenerationException(
        response.status,
        "STREAM_PROTOCOL_ERROR",
        "Governed generation did not return an SSE stream",
      );
    }

    reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    const parseFrames = (flush: boolean): GovernedStreamEvent[] => {
      const frames: string[] = [];
      const delimiter = /\r\n\r\n|\r\r|\n\n/g;
      let frameStart = 0;
      let match: RegExpExecArray | null;
      while ((match = delimiter.exec(buffer)) !== null) {
        frames.push(buffer.slice(frameStart, match.index));
        frameStart = match.index + match[0].length;
      }
      buffer = buffer.slice(frameStart);
      if (flush && buffer.length > 0) {
        frames.push(buffer);
        buffer = "";
      }
      const events: GovernedStreamEvent[] = [];
      for (const frame of frames) {
        const event = parseSseFrame(frame, response.status);
        if (event) events.push(event);
      }
      return events;
    };

    while (true) {
      const { done, value } = await reader.read();
      if (done) {
        buffer += decoder.decode();
        if (buffer.length > 0) buffer += "\n\n";
      } else {
        buffer += decoder.decode(value, { stream: true });
      }

      for (const event of parseFrames(done)) {
        if (event.event === "complete") terminalSeen = true;
        yield event;
        if (event.event === "complete") return;
      }
      if (done) break;
    }

    throw new GovernedGenerationException(
      response.status,
      "STREAM_TERMINATED",
      "Governed generation stream ended without a terminal event",
    );
  } catch (error) {
    if (error instanceof GovernedGenerationException) throw error;
    if (controller.signal.aborted) {
      throw new GovernedGenerationException(
        null,
        "TRANSPORT_TIMEOUT",
        "Governed generation stream timed out",
      );
    }
    throw new GovernedGenerationException(
      null,
      "TRANSPORT_ERROR",
      "Governed generation stream failed",
      error,
    );
  } finally {
    clearTimeout(timer);
    if (reader) {
      if (!terminalSeen) await reader.cancel();
      reader.releaseLock();
    }
  }
}

function buildBody(
  prompt: string,
  options: GenerateGovernedOptions,
): GovernedGenerateWireRequest {
  if (typeof prompt !== "string" || !prompt.trim() || prompt.length > 10_000) {
    throw new Error("prompt must be a non-blank string of at most 10,000 characters");
  }
  if (!options.policyIds?.length && !options.bundleIds?.length) {
    throw new Error("Provide at least one of bundleIds or policyIds");
  }
  const maxRetries = options.maxRetries ?? 1;
  if (maxRetries !== 0 && maxRetries !== 1) {
    throw new Error("maxRetries must be 0 or 1");
  }

  const body: GovernedGenerateWireRequest = { prompt, max_retries: maxRetries };
  if (options.policyIds?.length) body.policy_ids = [...options.policyIds];
  if (options.bundleIds?.length) body.bundle_ids = [...options.bundleIds];
  if (options.generation) {
    const { temperature, maxTokens } = options.generation;
    if (
      temperature !== undefined
      && (!Number.isFinite(temperature) || temperature < 0 || temperature > 2)
    ) {
      throw new Error("generation.temperature must be between 0 and 2");
    }
    if (
      maxTokens !== undefined
      && (!Number.isInteger(maxTokens) || maxTokens < 1 || maxTokens > 4096)
    ) {
      throw new Error("generation.maxTokens must be an integer between 1 and 4096");
    }
    body.generation = {
      ...(temperature !== undefined ? { temperature } : {}),
      ...(maxTokens !== undefined ? { max_tokens: maxTokens } : {}),
    };
  }
  return body;
}

function buildHeaders(
  config: GovernedTransportConfig,
  accept: "application/json" | "text/event-stream",
): Record<string, string> {
  const headers: Record<string, string> = {
    Authorization: `Bearer ${config.apiKey}`,
    Accept: accept,
    "Content-Type": "application/json",
  };
  if (config.providerKey) {
    headers["X-Provider-Key"] = config.providerKey;
    if (config.providerName) headers["X-Provider"] = config.providerName;
  }
  return headers;
}

async function decodeJson(response: Response): Promise<UnknownRecord> {
  let payload: unknown;
  try {
    payload = await response.json();
  } catch (error) {
    throw new GovernedGenerationException(
      response.status,
      "PARSE_ERROR",
      "Failed to parse governed generation response",
      error,
    );
  }
  if (!isRecord(payload)) {
    throw new GovernedGenerationException(
      response.status,
      "PARSE_ERROR",
      "Governed generation response must be a JSON object",
    );
  }
  return payload;
}

function throwApiError(status: number, payload: UnknownRecord): never {
  const error = isRecord(payload.error) ? payload.error : {};
  const code = typeof error.code === "string" ? error.code : "UNKNOWN_ERROR";
  const message = typeof error.message === "string"
    ? error.message
    : `Governed generation failed with HTTP ${status}`;
  const details = error.details ?? payload.data;
  if (status === 422 && code === "GOVERNED_OUTPUT_BLOCKED") {
    throw new GovernanceDeniedException(
      message,
      parseBlockedData(payload.data, status),
    );
  }
  throw new GovernedGenerationException(status, code, message, details);
}

function parseSseFrame(
  frame: string,
  status: number,
): GovernedStreamEvent | null {
  let eventName: string | undefined;
  const dataLines: string[] = [];
  for (const line of frame.split(/\r\n|\r|\n/)) {
    if (line.startsWith(":")) continue;
    const colon = line.indexOf(":");
    const field = colon === -1 ? line : line.slice(0, colon);
    let value = colon === -1 ? "" : line.slice(colon + 1);
    if (value.startsWith(" ")) value = value.slice(1);
    if (field === "event") eventName = value;
    if (field === "data") dataLines.push(value);
  }

  if (!eventName || !["status", "heartbeat", "complete", "blocked", "error"].includes(eventName)) {
    return null;
  }
  if (!dataLines.length) {
    throw new GovernedGenerationException(
      status,
      "STREAM_PARSE_ERROR",
      `SSE event ${eventName} has no data`,
    );
  }

  let payload: unknown;
  try {
    payload = JSON.parse(dataLines.join("\n"));
  } catch (error) {
    throw new GovernedGenerationException(
      status,
      "STREAM_PARSE_ERROR",
      `Failed to parse governed generation SSE event: ${eventName}`,
      error,
    );
  }
  if (!isRecord(payload)) {
    throw new GovernedGenerationException(
      status,
      "STREAM_PARSE_ERROR",
      `Governed generation SSE event ${eventName} must contain an object`,
    );
  }

  try {
    if (eventName === "status") {
      const stage = stringField(payload, "stage");
      if (!STATUS_STAGES.has(stage as GovernedStatusStage)) {
        throw new TypeError("Invalid governed status stage");
      }
      const event: GovernedStatusEvent = {
        event: "status",
        data: {
          stage: stage as GovernedStatusStage,
          attempt: integerField(payload, "attempt"),
          ...(payload.violations !== undefined
            ? { violations: integerValue(payload.violations, "violations") }
            : {}),
        },
      };
      return event;
    }
    if (eventName === "heartbeat") {
      const event: GovernedHeartbeatEvent = {
        event: "heartbeat",
        data: { timestamp: stringField(payload, "timestamp") },
      };
      return event;
    }
    if (eventName === "complete") {
      if (payload.success !== true) throw new TypeError("Invalid complete event");
      const event: GovernedCompleteEvent = {
        event: "complete",
        data: { success: true, data: parseCompleteData(payload.data, status) },
      };
      return event;
    }

    const error = recordField(payload, "error");
    const code = stringField(error, "code");
    const message = stringField(error, "message");
    const logicalStatus = integerField(error, "http_status");
    if (eventName === "blocked") {
      if (code !== "GOVERNED_OUTPUT_BLOCKED" || logicalStatus !== 422) {
        throw new TypeError("Invalid blocked terminal event");
      }
      throw new GovernanceDeniedException(
        message,
        parseBlockedData(payload.data, logicalStatus),
      );
    }
    throw new GovernedGenerationException(
      logicalStatus,
      code,
      message,
      payload.data,
    );
  } catch (error) {
    if (error instanceof GovernedGenerationException) throw error;
    throw new GovernedGenerationException(
      status,
      "STREAM_PARSE_ERROR",
      `Invalid governed generation SSE event: ${eventName}`,
      error,
    );
  }
}

function parseCompleteData(value: unknown, status: number): GovernedCompleteData {
  try {
    const data = recordValue(value, "data");
    return {
      content: stringField(data, "content"),
      provider: stringField(data, "provider"),
      model: stringField(data, "model"),
      ...(data.usage !== undefined ? { usage: parseUsage(data.usage) } : {}),
      attempts: integerField(data, "attempts"),
      attempt_metadata: parseAttempts(data.attempt_metadata),
      evaluation: parseEvaluation(data.evaluation),
      accounting: parseAccounting(data.accounting),
      total_duration_ms: integerField(data, "total_duration_ms"),
    };
  } catch (error) {
    if (error instanceof GovernedGenerationException) throw error;
    throw new GovernedGenerationException(
      status,
      "PARSE_ERROR",
      "Invalid governed generation completion payload",
      error,
    );
  }
}

function parseBlockedData(value: unknown, status: number): GovernedBlockedData {
  try {
    const data = recordValue(value, "data");
    return {
      attempts: integerField(data, "attempts"),
      attempt_metadata: parseAttempts(data.attempt_metadata),
      evaluation: parseEvaluation(data.evaluation),
      accounting: parseAccounting(data.accounting),
      total_duration_ms: integerField(data, "total_duration_ms"),
    };
  } catch (error) {
    if (error instanceof GovernedGenerationException) throw error;
    throw new GovernedGenerationException(
      status,
      "PARSE_ERROR",
      "Invalid governed generation blocked payload",
      error,
    );
  }
}

function parseUsage(value: unknown): GovernedTokenUsage {
  const usage = recordValue(value, "usage");
  return {
    prompt_tokens: integerField(usage, "prompt_tokens"),
    completion_tokens: integerField(usage, "completion_tokens"),
    total_tokens: integerField(usage, "total_tokens"),
  };
}

function parseAttempts(value: unknown): GovernedAttemptMetadata[] {
  if (!Array.isArray(value)) throw new TypeError("attempt_metadata must be an array");
  return value.map(item => {
    const attempt = recordValue(item, "attempt_metadata item");
    return {
      attempt: integerField(attempt, "attempt"),
      provider: stringField(attempt, "provider"),
      model: stringField(attempt, "model"),
      generation_duration_ms: integerField(attempt, "generation_duration_ms"),
      evaluation_duration_ms: integerField(attempt, "evaluation_duration_ms"),
      policies_evaluated: integerField(attempt, "policies_evaluated"),
      allowed: booleanField(attempt, "allowed"),
      ...(attempt.usage !== undefined ? { usage: parseUsage(attempt.usage) } : {}),
    };
  });
}

function parseAccounting(value: unknown): GovernedAccounting {
  const accounting = recordValue(value, "accounting");
  return {
    generation_attempts: integerField(accounting, "generation_attempts"),
    evaluation_batches: integerField(accounting, "evaluation_batches"),
    policy_evaluations: integerField(accounting, "policy_evaluations"),
  };
}

function parseEvaluation(value: unknown): GovernedEvaluationSummary {
  const evaluation = recordValue(value, "evaluation");
  const receiptId = evaluation.receipt_id;
  if (receiptId !== undefined && typeof receiptId !== "string") {
    throw new TypeError("receipt_id must be a string");
  }
  return {
    allowed: booleanField(evaluation, "allowed"),
    policy_ids: stringArrayField(evaluation, "policy_ids"),
    policies_evaluated: integerField(evaluation, "policies_evaluated"),
    policies_passed: integerField(evaluation, "policies_passed"),
    policies_failed: integerField(evaluation, "policies_failed"),
    policies_errored: integerField(evaluation, "policies_errored"),
    violation_count: integerField(evaluation, "violation_count"),
    statutory_anchors: stringArrayField(evaluation, "statutory_anchors"),
    ...(receiptId !== undefined ? { receipt_id: receiptId } : {}),
  };
}

function isRecord(value: unknown): value is UnknownRecord {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function recordValue(value: unknown, name: string): UnknownRecord {
  if (!isRecord(value)) throw new TypeError(`${name} must be an object`);
  return value;
}

function recordField(record: UnknownRecord, key: string): UnknownRecord {
  return recordValue(record[key], key);
}

function stringField(record: UnknownRecord, key: string): string {
  const value = record[key];
  if (typeof value !== "string") throw new TypeError(`${key} must be a string`);
  return value;
}

function integerValue(value: unknown, name: string): number {
  if (!Number.isInteger(value)) throw new TypeError(`${name} must be an integer`);
  return value as number;
}

function integerField(record: UnknownRecord, key: string): number {
  return integerValue(record[key], key);
}

function booleanField(record: UnknownRecord, key: string): boolean {
  const value = record[key];
  if (typeof value !== "boolean") throw new TypeError(`${key} must be a boolean`);
  return value;
}

function stringArrayField(record: UnknownRecord, key: string): string[] {
  const value = record[key];
  if (!Array.isArray(value) || value.some(item => typeof item !== "string")) {
    throw new TypeError(`${key} must be an array of strings`);
  }
  return value as string[];
}
